Level 1

Andreysergeev91 opened the pull request:
В проекте две сущности: пользователя и статьи (user и article). Созданы схема и модель для каждой.
Созданы роуты и контроллеры:   GET /users/me, GET /articles, POST /articles, DELETE /articles/articleId.
Создан контроллер для каждого роута. Роуты защищены авторизацией.
В API реализованы роуты для регистрации и логина.
Реализовано логгирование.
Создан облачный сервер, развернуто API;
Создан домен и прикреплен к серверу;
Добавлено шифрование, к серверу можно обратиться по https;
На сервер добавлен .env файл с секретным ключом;

Andreysergeev91 committed:
Add model and schema 'user'

Add Mongoose model and schema 'user'

Andreysergeev91 committed:
Add model and schema 'article'

Add Mongoose  model and schema 'article'

Andreysergeev91 committed:
Add user controllers

Add user controllers for creation and logging users,
also for getting user info

Andreysergeev91 committed:
Add article controllers

Add article controllers for creation, review and removing users
articles

Andreysergeev91 committed:
Add middleware 'auth'

Add middleware 'auth' for users authentication

Andreysergeev91 committed:
Add custom error classes

Add custom error classes by inheritance of built-in error class
for set up appropriate status code

Andreysergeev91 committed:
Add users routes

Andreysergeev91 committed:
Add articles routes

Andreysergeev91 committed:
Add middleware 'logger.js '

Add middleware logger.js for traceability of errors and requests

Andreysergeev91 committed:
Add middleware ratelimiter

Andreysergeev91 committed:
Add middleware 'main-error-handler '

Add middleware main-error-handler for centralized error handling

Andreysergeev91 committed:
Add request validation for user controllers

Add request validation for data in req.body

Andreysergeev91 committed:
Add controller "getStatusCodeByError"

Add additional controller for determination of error status code

Andreysergeev91 committed:
Add .editorconfig

Andreysergeev91 committed:
Add .eslintrc

Andreysergeev91 committed:
Add .gitignore

Andreysergeev91 committed:
Add package.json and package-lock.json

Andreysergeev91 committed:
Add app.js

Andreysergeev91 committed:
Relocate user routes from app.js

Relocate user routes from app.js and add auth middleware
to routes

Andreysergeev91 committed:
Relocate request validation in a separate file

Relocate request validation in a separate file and
add auth middleware to article routes

Andreysergeev91 committed:
Relocate all request validation

Relocate all request validation in a separate file

Andreysergeev91 committed:
Add main router "index.js"

Add main router "index.js"
for routing user and article routes

Andreysergeev91 committed:
Relocate users routes

Relocate users routes and error-path middleware
 in a separate file

Andreysergeev91 committed:
Update README.md

Andreysergeev91 committed:
Update README.md

Andreysergeev91 committed:
Add select() to deleteArticle controller

Add select() to deleteArticle controller
for searching article with article.owner

Boortcore commented:
Добрый день!
На данный момент невозможно создать пользователя с валидным паролем. Проект имеет ограничение на содержимое пароля, которое не предусмотрено заданием. 
Пожалуйста, исправьте валидацию пароля. 
![image](https://user-images.githubusercontent.com/17438227/85927774-eee59680-b8b0-11ea-9d08-953ca417c521.png)

Andreysergeev91 committed:
Remove unnecessary validation

Remove unnecessary password and keyword validation

Andreysergeev91 committed:
Remove unnecessary validation

Remove unnecessary keyword validation

Andreysergeev91 committed:
Fix eslint error and remove validation

Fix eslint error in userSchema.statics
through assigning of function name and
remove password validation

Andreysergeev91 committed:
Fix eslint error

Andreysergeev91 committed:
Fix eslint error

Andreysergeev91 committed:
Fix eslint error

Andreysergeev91 committed:
Change connection string and fix eslint error

Change connection string and fix eslint error
by removing of console.error.

Andreysergeev91 commented:
Добрый день!
Спасибо за проверку.
Исправил валидацию пароля и ключевого слова.
Исправил ошибки eslint'a.
Изменил адрес БД в production режиме.

Boortcore commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [x] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
  - [x] в проекте есть package.json;
  - [x] .editorconfig;
  - [x] .eslintrc, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
  - [x] .gitignore.
  - [x] в .eslintrc добавлено исключение для _id _;
- [x] <!--3.64--> Нет ошибок линтера.
- [x] <!--3.64--> В разделе scripts файла package.json:
  - [x] есть команда npm run start, которая запускает сервер на localhost:3000;
  - [x] и npm run dev, которая запускает сервер на localhost:3000 с хот релоудом.
- [x] <!--12.57--> Когда все зависимости установлены, приложение запускается командой npm run dev без ошибок.
- [x] <!--25.45--> Описанные роуты работают корректно:
  - [x] запрос на GET /users/me возвращает информацию о пользователе (email и имя);
  - [x] GET /articles — все сохранённые статьи;
  - [x] POST /articles создаёт статью с переданными в теле данными;
  - [x] DELETE /articles/articleId удаляет статью по _id _; 
  - [x] POST /signup создаёт пользователя с переданными в теле данными;
  - [x] POST /signin возвращает JWT, если в теле запроса переданы правильные почта и пароль.
- [x] <!--3.64--> Все роуты, кроме /signin и /signup, защищены авторизацией.
- [x] <!--3.64--> Роуты пользователей и роуты статей описаны в отдельных файлах.
- [ ] <!--3.64--> Ошибки API обрабатываются:
  - [x] если в запросе что-то не так, сервер возвращает ответ с ошибкой и её статусом;
  - [x] асинхронные обработчики завершены блоком catch;
  - [ ] API не возвращает стандартных ошибок базы данных или Node.js.

При неудачной попытке залогиниться возвращается данная ошибка.
![image](https://user-images.githubusercontent.com/17438227/86000215-e9ee2780-ba15-11ea-9da4-3c9c7c7fcab2.png)

- [x] <!--3.64--> В production-режиме адрес базы данных берётся из process.env.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
  - [x] пароль хранится в зашифрованном виде;
  - [x] API не возвращает хеш пароля клиенту.
- [x] <!--3.64--> Данные валидируются перед добавлением в базу.
- [x] <!--3.64--> Пользователь не может удалить сохранённую карточку из профиля другого пользователя.
- [x] <!--3.64--> К серверу можно обратиться и по http, и по https, обращаясь к домену, указанному в README.md.
- [x] <!--3.64--> Правильно реализовано хранение секретного ключа для создания JWT:
  - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
  - [x] в режиме разработки (при process.env.NODE_ENV !== 'production') код запускается и работает и без .env файла.

### Хорошие практики
- [x] <!--1.5--> Асинхронные операции реализованы промисами или async/await.
- [ ] <!--1.5--> Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры, проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации.

При создании статьи с невалидным свойством link или image запрос не должен уходить на контроллер и обрабатываться базой. Такой запрос должен быть прерван пакетом celebrate, после чего сообщение об ошибке отправляется на клиент
Joi.string().required() пропускает любую строку, которая может не быть валидной ссылкой. Для правильной валидации ссылок необходимо использовать регулярные выражения или метод custom ![image](https://user-images.githubusercontent.com/17438227/83885018-732f7880-a74e-11ea-905e-28d6e8ef00d1.png)
В качестве валидации ссылок предлагаю использовать метод `custom` https://github.com/hapijs/joi/blob/master/API.md#anycustommethod-description

- [x] <!--1.5--> Настроено логгирование:
  - [x] запросы и ответы записываются в файл request.log;
  - [x] ошибки записываются в файл error.log;
  - [x] файлы логов не добавляются в репозиторий.
- [x] <!--1.5--> Ошибки обрабатываются централизованным обработчиком.
- [x] <!--1.5--> Централизованный обработчик ошибок описан в отдельном модуле.
- [x] <!--1.5--> В случае ошибки API возвращает статус ответа, который соответствует типу ошибки.
- [x] <!--1.5--> К серверу можно обратиться по https по адресу домена, указанному в README.md.
- [x] <!--1.5--> API приложения располагается на домене вида: name.zone/api, а не просто name.zone
- [x] <!--1.5--> Все роуты подключены в файле index.js, который находится в папке routes. Оттуда единый роут подключается в файле app.js.

### Рекомендации
- [x] <!--0.83--> Для ошибок API созданы классы, расширяющие конструктор Error.
- [x] <!--0.83--> Используется модуль Helmet для установки заголовков, связанных с безопасностью.
- [ ] <!--0.83--> Конфигурация и константы в отдельных файлах:
  - [ ] адрес Mongo-сервера в режиме разработки вынесен в отдельный конфигурационный файл;
  - [ ] секретный ключ для JWT в режиме разработки вынесен в отдельный конфигурационный файл;
  - [ ] константы приложения — сообщения ответов и ошибок — вынесены в отдельный файл с константами.
- [x] <!--0.83--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено.
- [x] <!--0.83--> rate limiter сконфигурирован в отдельном файле и импортируется в app.js
- [x] <!--0.83--> API размещено на отдельном поддомене, например: api.news-explorer.tk.

## Количество баллов: 94

## Доп. Комментарии:
- API возвращает хеш пароля при создании пользователя. Нужно исправить.

Andreysergeev91 committed:
Add module 'cors'

Andreysergeev91 committed:
Change CORS options

Andreysergeev91 committed:
Change CORS options

Andreysergeev91 committed:
Add signOut controller, change createUser

Andreysergeev91 committed:
Add signOut route

Andreysergeev91 committed:
Change CORS options

Andreysergeev91 committed:
Add sameSite option to res.cookie

Andreysergeev91 committed:
Change sameSite option

Andreysergeev91 committed:
Change res.cookie options

Andreysergeev91 committed:
Change sameSite option in login

Andreysergeev91 committed:
Change SameSite options

Andreysergeev91 committed:
Change SameSite options
